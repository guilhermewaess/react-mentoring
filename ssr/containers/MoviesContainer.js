// import { parse, stringify } from 'query-string';
import * as React from 'react';
import { connect } from 'react-redux';
import { getMovies } from './../store/actions';
import MoviesList from './../components/MoviesList';
import Header from './../components/MoviesHeader';
import FilterResults from './../components/FilterResults';

const mapStateToProps = state => {
  return {
    movies: state.movies,
  };
};

const mapActionsToProps = dispatch => {
  return {
    getMovies: filter => dispatch(getMovies(filter)),
  };
};

class MoviesContainer extends React.Component {
  state;
  historySubscription;

  constructor(props) {
    super(props);
    this.state = {
      filter: {
        search: '',
        searchBy: 'title',
        sortBy: 'release_date',
        sortOrder: 'asc',
      },
    };
    // this.historySubscription = this.props.history.listen(
    //   this.search.bind(this),
    // );
  }

  // componentDidMount() {
  //   const queryParams = parse(this.props.location.search);
  //   this.updateFilter(queryParams);
  // }

  // componentWillUnmount() {
  //   this.historySubscription();
  // }

  search() {
    this.props.getMovies(this.state.filter);
  }

  onSearchHandler(event) {
    event.preventDefault();
    this.updateFilter();
  }

  onFilterChangeHandler(shouldSearch, event) {
    event.stopPropagation();
    const filter = {
      ...this.state.filter,
      [event.target.dataset.filterField]: event.target.value,
    };
    this.setState({ filter }, () => {
      if (shouldSearch) {
        this.updateFilter();
      }
    });
  }

  updateFilter(queryParams) {
    const filter = {
      ...this.state.filter,
      ...queryParams,
    };

    this.setState({ filter }, this.updateQueryStringParams);
  }

  updateQueryStringParams() {
    this.props.history.push({
      pathname: '/movies',
      search: stringify(this.state.filter),
    });
  }

  render() {
    const onSearch = this.onSearchHandler.bind(this);
    const onFilterChange = this.onFilterChangeHandler.bind(this, false);
    const onSortOrderByChange = this.onFilterChangeHandler.bind(this, true);
    return (
      <div className="w-100 h-100">
        {JSON.stringify(this.state)}
        <Header
          handleFilterChange={onFilterChange}
          onSearch={onSearch}
          filter={this.state.filter}
        />
        <FilterResults
          movies={this.props.movies}
          onFilterChange={onSortOrderByChange}
          filter={this.state.filter}
        />
        <MoviesList movies={this.props.movies} />
      </div>
    );
  }
}

export default connect(mapStateToProps, mapActionsToProps)(MoviesContainer);
