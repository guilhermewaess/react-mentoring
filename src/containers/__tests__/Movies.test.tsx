import { shallow, ShallowWrapper } from 'enzyme';
import * as React from 'react';
import * as configureStore from 'redux-mock-store';
import MoviesContainer from './../Movies';

// jest.unmock('redux-mock-store');

describe('MoviesContainer', () => {
  let component: ShallowWrapper<any, any>;
  let instance: any;
  let props: any;
  let historySubscriptionCallback: any;
  let store: any;

  const helper = {
    createProps() {
      historySubscriptionCallback = jest.fn();
      const propsObject = {
        history: {
          listen: jest.fn(),
          push: jest.fn(),
        },
        location: { search: '?search=abc' },
        getMovies: jest.fn(),
        movies: [],
      };

      propsObject.history.listen.mockReturnValue(historySubscriptionCallback);

      return propsObject;
    },
  };

  beforeEach(() => {
    props = helper.createProps();
    store = configureStore()({});
    component = shallow(<MoviesContainer store={store} {...props} />).dive();
    instance = component.instance();
  });

  describe('when construct', () => {
    it('should subscribe history with search function', () => {
      expect(instance.historySubscription).toEqual(historySubscriptionCallback);
    });
  });

  describe('when call componentDidMount', () => {
    beforeEach(() => {
      instance.componentDidMount();
    });
    it('should have state updated with query params', () => {
      expect(component.state()).toEqual({
        filter: {
          search: 'abc',
          searchBy: 'title',
          sortBy: 'release_date',
          sortOrder: 'asc',
        },
      });
    });
    it('should update queryStrings with new filter', () => {
      expect(props.history.push).toHaveBeenCalledWith({
        pathname: '/movies',
        search: 'search=abc&searchBy=title&sortBy=release_date&sortOrder=asc',
      });
    });
  });

  describe('when call componentWillUnmount', () => {
    beforeEach(() => {
      instance.componentWillUnmount();
    });
    it('should execute historySubscriptions', () => {
      expect(historySubscriptionCallback).toHaveBeenCalledTimes(1);
    });
  });

  describe('when call search', () => {
    beforeEach(() => {
      instance.search();
    });
    it('should call getMovies with current filter', () => {
      expect(props.getMovies).toHaveBeenCalledWith(component.state().filter);
    });
  });
});
