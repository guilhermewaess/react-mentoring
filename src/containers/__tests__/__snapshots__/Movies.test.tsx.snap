// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movies when construct should render 1`] = `
<div
  className="w-100 h-100"
>
  <Header />
  <FilterResults
    movies={
      Array [
        Object {
          "budget": 0,
          "genres": Array [
            "Action",
            "Adventure",
            "Science Fiction",
          ],
          "id": 447365,
          "overview": "The third film based on Marvel's Guardians of the Galaxy.",
          "poster_path": "https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg",
          "release_date": "2020-05-01",
          "revenue": 0,
          "runtime": null,
          "tagline": "",
          "title": "Guardians of the Galaxy Vol. 3",
          "vote_average": 0,
          "vote_count": 9,
        },
        Object {
          "budget": 0,
          "genres": Array [
            "Action",
            "Adventure",
            "Science Fiction",
          ],
          "id": 447366,
          "overview": "The third film based on Marvel's Guardians of the Galaxy.",
          "poster_path": "https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg",
          "release_date": "2020-05-01",
          "revenue": 0,
          "runtime": null,
          "tagline": "",
          "title": "Guardians of the Galaxy Vol. 3",
          "vote_average": 0,
          "vote_count": 9,
        },
        Object {
          "budget": 0,
          "genres": Array [
            "Action",
            "Adventure",
            "Science Fiction",
          ],
          "id": 447367,
          "overview": "The third film based on Marvel's Guardians of the Galaxy.",
          "poster_path": "https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg",
          "release_date": "2020-05-01",
          "revenue": 0,
          "runtime": null,
          "tagline": "",
          "title": "Guardians of the Galaxy Vol. 3",
          "vote_average": 0,
          "vote_count": 9,
        },
        Object {
          "budget": 0,
          "genres": Array [
            "Action",
            "Adventure",
            "Science Fiction",
          ],
          "id": 447368,
          "overview": "The third film based on Marvel's Guardians of the Galaxy.",
          "poster_path": "https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg",
          "release_date": "2020-05-01",
          "revenue": 0,
          "runtime": null,
          "tagline": "",
          "title": "Guardians of the Galaxy Vol. 3",
          "vote_average": 0,
          "vote_count": 9,
        },
      ]
    }
  />
  <Component
    movies={
      Array [
        Object {
          "budget": 0,
          "genres": Array [
            "Action",
            "Adventure",
            "Science Fiction",
          ],
          "id": 447365,
          "overview": "The third film based on Marvel's Guardians of the Galaxy.",
          "poster_path": "https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg",
          "release_date": "2020-05-01",
          "revenue": 0,
          "runtime": null,
          "tagline": "",
          "title": "Guardians of the Galaxy Vol. 3",
          "vote_average": 0,
          "vote_count": 9,
        },
        Object {
          "budget": 0,
          "genres": Array [
            "Action",
            "Adventure",
            "Science Fiction",
          ],
          "id": 447366,
          "overview": "The third film based on Marvel's Guardians of the Galaxy.",
          "poster_path": "https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg",
          "release_date": "2020-05-01",
          "revenue": 0,
          "runtime": null,
          "tagline": "",
          "title": "Guardians of the Galaxy Vol. 3",
          "vote_average": 0,
          "vote_count": 9,
        },
        Object {
          "budget": 0,
          "genres": Array [
            "Action",
            "Adventure",
            "Science Fiction",
          ],
          "id": 447367,
          "overview": "The third film based on Marvel's Guardians of the Galaxy.",
          "poster_path": "https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg",
          "release_date": "2020-05-01",
          "revenue": 0,
          "runtime": null,
          "tagline": "",
          "title": "Guardians of the Galaxy Vol. 3",
          "vote_average": 0,
          "vote_count": 9,
        },
        Object {
          "budget": 0,
          "genres": Array [
            "Action",
            "Adventure",
            "Science Fiction",
          ],
          "id": 447368,
          "overview": "The third film based on Marvel's Guardians of the Galaxy.",
          "poster_path": "https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg",
          "release_date": "2020-05-01",
          "revenue": 0,
          "runtime": null,
          "tagline": "",
          "title": "Guardians of the Galaxy Vol. 3",
          "vote_average": 0,
          "vote_count": 9,
        },
      ]
    }
  />
</div>
`;
